ouroboros-consensus-diffusion/test/consensus-test/Test/Consensus/Genesis/Tests/Uniform.hs

```diff
@@ -0,0 +1,277 @@
+  counterexample killedPeers $
+  -- We require the honest chain to fit a Genesis window, because otherwise its tip may suggest
+  -- to the governor that the density is too low.
+  longerThanGenesisWindow ==>
+  conjoin [
+    counterexample "The honest peer was disconnected" (not (HonestPeer `elem` killed)),
+    counterexample ("The immutable tip is not honest: " ++ show immutableTip) $
+    property (isHonest immutableTipHash),
+    immutableTipIsRecent
+  ]
+  where
+    advCount = Map.size (others schedule)
+
+    immutableTipIsRecent =
+      counterexample ("Age of the immutable tip: " ++ show immutableTipAge) $
+      immutableTipAge `le` s + fromIntegral d + 1
+
+    SlotNo immutableTipAge = case (honestTipSlot, immutableTipSlot) of
+      (At h, At i)   -> h - i
+      (At h, Origin) -> h
+      _              -> 0
+
+    isOrigin = null
+
+    isHonest = all (0 ==)
+
+    immutableTipHash = simpleHash (AF.anchorToHash immutableTip)
+
+    immutableTip = AF.anchor svSelectedChain
+
+    immutableTipSlot = AF.anchorToSlotNo (AF.anchor svSelectedChain)
+
+    disconnected =
+      printf "disconnected %.1f%% of adversaries" disconnectedPercent
+
+    disconnectedPercent :: Double
+    disconnectedPercent =
+      100 * fromIntegral (length killed) / fromIntegral advCount
+
+    killedPeers = case killed of
+      [] -> "No peers were killed"
+      peers -> "Some peers were killed: " ++ intercalate ", " (condense <$> peers)
+
+    honestTipSlot = At $ tbSlot $ snd $ last $ mapMaybe fromBlockPoint $ value $ honest schedule
+
+    GenesisTest {gtBlockTree, gtGenesisWindow = GenesisWindow s, gtDelay = Delta d} = genesisTest
+
+    Classifiers {genesisWindowAfterIntersection, longerThanGenesisWindow} = classifiers genesisTest
+
+fromBlockPoint :: (DiffTime, SchedulePoint) -> Maybe (DiffTime, TestBlock)
+fromBlockPoint (t, ScheduleBlockPoint bp) = Just (t, bp)
+fromBlockPoint _                          = Nothing
+
+-- | Tests that the immutable tip is not delayed and stays honest with the
+-- adversarial peers serving adversarial branches.
+prop_serveAdversarialBranches :: QC.Gen QC.Property
+prop_serveAdversarialBranches = QC.expectFailure <$> do
+  genesisTest <- genChains (QC.choose (1, 4))
+  schedulePoints <- genUniformSchedulePoints genesisTest
+  pure $
+    runSimOrThrow $
+    runTest schedulerConfig genesisTest (fromSchedulePoints schedulePoints) $
+    exceptionCounterexample $
+    makeProperty genesisTest schedulePoints
+
+  where
+    schedulerConfig = (noTimeoutsSchedulerConfig scheduleConfig) {scTraceState = False, scTrace = False}
+
+    scheduleConfig = defaultPointScheduleConfig
+
+genUniformSchedulePoints :: GenesisTest -> QC.Gen (Peers PeerSchedule)
+genUniformSchedulePoints gt = stToGen (uniformPoints (gtBlockTree gt))
+
+-- Note [Leashing attacks]
+--
+-- A leashing attack would be rehearsed by a point schedule meeting either of
```

Comment:
I don't know if `rehearsed` means any of these:

```suggestion
-- A leashing attack would be exercised by a point schedule meeting either of
```

```suggestion
-- A leashing attack would be simulated by a point schedule meeting either of
```

```suggestion
-- A leashing attack would be modelled by a point schedule meeting either of
```

Reply:
I went for “successfully conducted” in d96aee77387eccb196522b4e5b3df1b61518440c.